File: frontend/src/types/cashflow.ts
export type CurveType =
  | 'flat'
  | 'bell'
  | 'front_loaded'
  | 'back_loaded'
  | 's_curve'
  | 'shoot_proportional'
  | 'milestone';

export type PhaseAssignment =
  | 'prep'
  | 'production'
  | 'post'
  | 'delivery'
  | 'full_span'
  | 'prep_and_production'
  | 'production_and_post';

export interface LineItemDistribution {
  budget_code: string;
  phase: PhaseAssignment;
  curve: CurveType;
  curve_params?: Record<string, number>;
  milestone_dates?: string[];
  milestone_amounts?: number[];
  auto_assigned: boolean;
}

export interface WeekColumn {
  week_number: number;
  week_commencing: string;
  phase_label: string;
  is_hiatus: boolean;
  shoot_days: number;
}

export interface CashflowRow {
  code: string;
  description: string;
  total: number;
  weekly_amounts: number[];
}

export interface CashflowOutput {
  title: string;
  weeks: WeekColumn[];
  rows: CashflowRow[];
  weekly_totals: number[];
  cumulative_totals: number[];
  grand_total: number;
}

File: backend/app/models/cashflow.py
from datetime import date

from pydantic import BaseModel

from app.models.budget import ParsedBudget
from app.models.distribution import LineItemDistribution
from app.models.production import ProductionParameters

CASH_INFLOW_LABELS = [
    "Broadcaster 1",
    "Broadcaster 2",
    "Broadcaster 3",
    "CMF Funding",
    "Provincial Tax Credit",
    "Federal Tax Credit",
    "Distribution Guarantee",
]


class WeekColumn(BaseModel):
    """A single week in the cashflow timeline."""

    week_number: int
    week_commencing: date
    phase_label: str
    is_hiatus: bool = False
    shoot_days: int = 0
    is_payroll_week: bool | None = None  # None if no payroll cycle set


class CashflowRow(BaseModel):
    """A single budget line item's weekly distribution."""

    code: str
    description: str
    total: float
    weekly_amounts: list[float]


class CashInflowRow(BaseModel):
    """A single cash inflow source with weekly amounts."""

    label: str
    weekly_amounts: list[float]


class CashflowOutput(BaseModel):
    """The complete cashflow ready for Excel generation."""

    title: str
    weeks: list[WeekColumn]
    rows: list[CashflowRow]
    weekly_totals: list[float]
    cumulative_totals: list[float]
    grand_total: float
    cash_inflows: list[CashInflowRow] = []
    inflow_weekly_totals: list[float] = []
    inflow_cumulative_totals: list[float] = []


class GenerateRequest(BaseModel):
    """Request body for cashflow generation endpoints."""

    budget: ParsedBudget
    parameters: ProductionParameters
    distributions: list[LineItemDistribution]
